semantic-ast/src/AST/GenerateSyntax.hs

```diff
@@ -62,76 +66,80 @@ getAllSymbols language = do
       let named = if t == 0 then Named else Anonymous
       pure (n, named)
 
+shapeParameterName, annParameterName :: Name
+shapeParameterName = mkName "f"
+annParameterName = mkName "a"
+
 -- Auto-generate Haskell datatypes for sums, products and leaf types
 syntaxDatatype :: Ptr TS.Language -> [(String, Named)] -> Datatype -> Q [Dec]
 syntaxDatatype language allSymbols datatype = skipDefined $ do
-  typeParameterName <- newName "a"
+  shapeFunctorKind <- [t| * -> * |]
+  let traversalInstances = mappend <$> makeStandaloneDerivings (conT name) <*> makeTraversalInstances (conT name)
+      glue a b c = a : b <> c
+      name = mkName nameStr
+      generatedDatatype cons = dataD (cxt []) name [kindedTV shapeParameterName shapeFunctorKind, plainTV annParameterName] Nothing cons [deriveStockClause, deriveAnyClassClause]
+      deriveStockClause = derivClause (Just StockStrategy) [conT ''Generic, conT ''Generic1]
+      deriveAnyClassClause = derivClause (Just AnyclassStrategy) [conT ''Traversable1 `appT` varT (mkName "someConstraint")]
+      deriveGN = derivClause (Just NewtypeStrategy) [conT ''TS.SymbolMatching]
   case datatype of
-    SumType (DatatypeName _) _ subtypes -> do
-      types' <- fieldTypesToNestedSum subtypes
-      let fieldName = mkName ("get" <> nameStr)
-      con <- recC name [TH.varBangType fieldName (TH.bangType strictness (pure types' `appT` varT typeParameterName))]
-      hasFieldInstance <- makeHasFieldInstance (conT name) (varT typeParameterName) (varE fieldName)
-      traversalInstances <- makeTraversalInstances (conT name)
-      pure
-        (  NewtypeD [] name [PlainTV typeParameterName] Nothing con [deriveGN, deriveStockClause, deriveAnyClassClause]
-        :  hasFieldInstance
-        <> traversalInstances)
-    ProductType (DatatypeName datatypeName) named children fields -> do
-      con <- ctorForProductType datatypeName typeParameterName children fields
-      symbolMatchingInstance <- symbolMatchingInstance allSymbols name named datatypeName
-      traversalInstances <- makeTraversalInstances (conT name)
-      pure
-        (  generatedDatatype name [con] typeParameterName
-        :  symbolMatchingInstance
-        <> traversalInstances)
+    SumType (DatatypeName _) _ subtypes ->
+      let types' = fieldTypesToNestedSum shapeParameterName subtypes
```

Comment:
We can't use the above defined `generatedDatatype` for the generation of sum types since sum types are newtypes (ex: `data JSONValue f a = JSONValue { getValue :: (Array f :+: String f :+: EtCetera f) a }`. This change was introduced [here](https://github.com/tree-sitter/haskell-tree-sitter/pull/214) to reduce the complexity of the original structure ([explained here](https://github.com/tree-sitter/haskell-tree-sitter/issues/210)). Due to this difference between sums and leaves, we can't resort to taking advantage of the factored out `generatedDatatype` function which produces a datatype, not a newtype.